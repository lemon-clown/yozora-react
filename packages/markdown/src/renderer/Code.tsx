import type { ICode } from '@yozora/ast'
import Code from '@yozora/react-code'
import React, { useContext } from 'react'
import { YozoraMarkdownContextType } from '../context/context'

export const YozoraCodeRenderer: React.FC<ICode> = props => {
  const { darken, preferLinenos, codeRunners } = useContext(YozoraMarkdownContextType)

  const { lang, meta } = props

  // Remove trailing line endings.
  const value: string = props.value.replace(/[\r\n]+$/, '')

  return (
    <Code
      lang={lang}
      value={value}
      meta={meta}
      runners={codeRunners}
      darken={darken}
      preferLinenos={preferLinenos}
    />
  )
}

YozoraCodeRenderer.displayName = 'YozoraCodeRenderer'
