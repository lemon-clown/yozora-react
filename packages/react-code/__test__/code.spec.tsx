import { render } from '@testing-library/react'
import React from 'react'
import Code from '../src'

describe('prop types', () => {
  test('render with custom className', () => {
    const code = 'let a = 1'
    const className = 'custom-code'
    const view = render(<Code className={className} value={code} />)
    expect(view.asFragment()).toMatchSnapshot()
  })
})

describe('snapshot', () => {
  const code = `
    () => {
      const a = 1;
      const b = 2;
      const c = a * a + b * b;
      return <span>Answer: {c}</span>
    }
  `.trim()
  const lang = 'jsx'

  test('live', () => {
    const view = render(<Code meta="live maxlines=4" lang={lang} value={code} />)
    expect(view.asFragment()).toMatchSnapshot()
  })

  test('embed', () => {
    const view = render(<Code meta="embed" lang={lang} value={code} />)
    expect(view.asFragment()).toMatchSnapshot()
  })

  test('literal', () => {
    const view = render(
      <Code
        meta={'literal {1-2,2-1,4} title="/home/demo/a.tsx" collapsed'}
        lang={lang}
        value={code}
      />,
    )
    expect(view.asFragment()).toMatchSnapshot()
  })
})
