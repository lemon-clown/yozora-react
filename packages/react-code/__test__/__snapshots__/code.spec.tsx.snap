// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot embed 1`] = `
<div
  class="yozora-code-embed css-1lkb549 yozora-code"
/>
`;

exports[`snapshot literal 1`] = `
<div
  class="yozora-code-literal css-xgjmtr yozora-code"
>
  <div
    class="css-11952oq"
  >
    <span
      class="css-9y6172"
    >
      <span
        class="css-t4q37b"
        title="close"
      />
      <span
        class="css-p28dfd"
        title="minimize"
      />
      <span
        class="css-6luv6p"
        title="maximize"
      />
    </span>
    <span
      class="css-1oku3ac"
      title="/home/demo/a.tsx"
    >
      /home/demo/a.tsx  | 6 lines.
    </span>
    <span
      class="css-5thvko"
    >
      <button
        aria-label="Copy to clipboard"
        class="css-jbwb7c"
        data-copy-status="waiting"
        type="button"
      >
        copy
      </button>
    </span>
  </div>
  <code
    class="css-ehogbo"
  >
    <pre>
      <div
        class="yozora-code-highlighter__content css-1k8659m prism-code language-jsx"
        style="color:#d4d4d4;background-color:none;line-height:1.6rem;max-height:0"
      >
        <div
          class="yozora-code-highlighter__linenos css-e3qifw"
          style="width:2.2em"
        >
          <div
            class="css-1jtfvys"
          >
            <span>
              1
            </span>
          </div>
          <div
            class="css-1jtfvys"
          >
            <span>
              2
            </span>
          </div>
          <div
            class="css-17fnzo2"
          >
            <span>
              3
            </span>
          </div>
          <div
            class="css-1jtfvys"
          >
            <span>
              4
            </span>
          </div>
          <div
            class="css-17fnzo2"
          >
            <span>
              5
            </span>
          </div>
          <div
            class="css-17fnzo2"
          >
            <span>
              6
            </span>
          </div>
        </div>
        <div
          class="yozora-code-highlighter__codes css-rv5jjl"
        >
          <div
            class="css-1ttfemb"
          >
            <div
              class="token-line css-2m46fv"
              style="color:#d4d4d4"
            >
              <span
                class="token keyword"
                style="color:rgb(86, 156, 214)"
              >
                function
              </span>
              <span
                class="token plain"
              >
                 
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                (
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                )
              </span>
              <span
                class="token plain"
              >
                 
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                {
              </span>
              <span
                class="token plain"
              />
            </div>
            <div
              class="token-line css-2m46fv"
              style="color:#d4d4d4"
            >
              <span
                class="token plain"
              >
                      
              </span>
              <span
                class="token keyword"
                style="color:rgb(86, 156, 214)"
              >
                const
              </span>
              <span
                class="token plain"
              >
                 a 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                =
              </span>
              <span
                class="token plain"
              >
                 
              </span>
              <span
                class="token number"
                style="color:rgb(181, 206, 168)"
              >
                1
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                ;
              </span>
              <span
                class="token plain"
              />
            </div>
            <div
              class="token-line css-7zjwmr"
              style="color:#d4d4d4"
            >
              <span
                class="token plain"
              >
                      
              </span>
              <span
                class="token keyword"
                style="color:rgb(86, 156, 214)"
              >
                const
              </span>
              <span
                class="token plain"
              >
                 b 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                =
              </span>
              <span
                class="token plain"
              >
                 
              </span>
              <span
                class="token number"
                style="color:rgb(181, 206, 168)"
              >
                2
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                ;
              </span>
              <span
                class="token plain"
              />
            </div>
            <div
              class="token-line css-2m46fv"
              style="color:#d4d4d4"
            >
              <span
                class="token plain"
              >
                      
              </span>
              <span
                class="token keyword"
                style="color:rgb(86, 156, 214)"
              >
                const
              </span>
              <span
                class="token plain"
              >
                 c 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                =
              </span>
              <span
                class="token plain"
              >
                 a 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                *
              </span>
              <span
                class="token plain"
              >
                 a 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                +
              </span>
              <span
                class="token plain"
              >
                 b 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                *
              </span>
              <span
                class="token plain"
              >
                 b
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                ;
              </span>
              <span
                class="token plain"
              />
            </div>
            <div
              class="token-line css-7zjwmr"
              style="color:#d4d4d4"
            >
              <span
                class="token plain"
              >
                      
              </span>
              <span
                class="token keyword"
                style="color:rgb(86, 156, 214)"
              >
                return
              </span>
              <span
                class="token plain"
              >
                 
              </span>
              <span
                class="token tag punctuation"
                style="color:rgb(106, 153, 85)"
              >
                &lt;
              </span>
              <span
                class="token tag"
                style="color:rgb(86, 156, 214)"
              >
                span
              </span>
              <span
                class="token tag punctuation"
                style="color:rgb(106, 153, 85)"
              >
                &gt;
              </span>
              <span
                class="token plain-text"
              >
                Answer: 
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                {
              </span>
              <span
                class="token plain"
              >
                c
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                }
              </span>
              <span
                class="token tag punctuation"
                style="color:rgb(106, 153, 85)"
              >
                &lt;/
              </span>
              <span
                class="token tag"
                style="color:rgb(86, 156, 214)"
              >
                span
              </span>
              <span
                class="token tag punctuation"
                style="color:rgb(106, 153, 85)"
              >
                &gt;
              </span>
              <span
                class="token plain"
              />
            </div>
            <div
              class="token-line css-7zjwmr"
              style="color:#d4d4d4"
            >
              <span
                class="token plain"
              >
                    
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                }
              </span>
            </div>
          </div>
        </div>
      </div>
    </pre>
  </code>
</div>
`;

exports[`snapshot live 1`] = `
<div
  class="yozora-code-live css-1wlwu1s yozora-code"
>
  <div
    class="css-11952oq"
  >
    <span
      class="css-9y6172"
    >
      <span
        class="css-t4q37b"
        title="close"
      />
      <span
        class="css-p28dfd"
        title="minimize"
      />
      <span
        class="css-6luv6p"
        title="maximize"
      />
    </span>
    <span
      class="css-1oku3ac"
      title=""
    >
       
    </span>
    <span
      class="css-5thvko"
    >
      <button
        aria-label="Copy to clipboard"
        class="css-jbwb7c"
        data-copy-status="waiting"
        type="button"
      >
        copy
      </button>
    </span>
  </div>
  <div
    class="css-1xlu443"
  >
    <div
      class="css-zquy5p"
    >
      <div
        class="yozora-code-editor css-kc9la2 yozora-code-editor"
        padding="10"
        style="line-height:1.6rem;tab-size:2"
      >
        <div
          class="css-6gcr58"
        >
          <div
            class="css-1tu59u4"
          />
          <textarea
            autocapitalize="off"
            autocomplete="off"
            autocorrect="off"
            class="yozora-code-editor__textarea css-16s63i0"
            spellcheck="false"
            style="tab-size:2;white-space:pre;padding:0.5rem 0.6em 0 12px"
          >
            function () {
      const a = 1;
      const b = 2;
      const c = a * a + b * b;
      return &lt;span&gt;Answer: {c}&lt;/span&gt;
    }
          </textarea>
        </div>
        <pre
          aria-hidden="true"
          class="css-14t25us"
          style="tab-size:2;white-space:pre"
        >
          <div
            class="yozora-code-highlighter__content css-1k8659m prism-code language-jsx"
            style="color:#d4d4d4;background-color:none;line-height:1.6rem;max-height:8rem;min-height:100%"
          >
            <div
              class="yozora-code-highlighter__linenos css-e3qifw"
              style="width:2.2em"
            >
              <div
                class="css-17fnzo2"
              >
                <span>
                  1
                </span>
              </div>
              <div
                class="css-17fnzo2"
              >
                <span>
                  2
                </span>
              </div>
              <div
                class="css-17fnzo2"
              >
                <span>
                  3
                </span>
              </div>
              <div
                class="css-17fnzo2"
              >
                <span>
                  4
                </span>
              </div>
              <div
                class="css-17fnzo2"
              >
                <span>
                  5
                </span>
              </div>
              <div
                class="css-17fnzo2"
              >
                <span>
                  6
                </span>
              </div>
            </div>
            <div
              class="yozora-code-highlighter__codes css-rv5jjl"
            >
              <div
                class="css-1ttfemb"
              >
                <div
                  class="token-line css-7zjwmr"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token keyword"
                    style="color:rgb(86, 156, 214)"
                  >
                    function
                  </span>
                  <span
                    class="token plain"
                  >
                     
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    (
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    )
                  </span>
                  <span
                    class="token plain"
                  >
                     
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    {
                  </span>
                  <span
                    class="token plain"
                  />
                </div>
                <div
                  class="token-line css-7zjwmr"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token plain"
                  >
                          
                  </span>
                  <span
                    class="token keyword"
                    style="color:rgb(86, 156, 214)"
                  >
                    const
                  </span>
                  <span
                    class="token plain"
                  >
                     a 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    =
                  </span>
                  <span
                    class="token plain"
                  >
                     
                  </span>
                  <span
                    class="token number"
                    style="color:rgb(181, 206, 168)"
                  >
                    1
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    ;
                  </span>
                  <span
                    class="token plain"
                  />
                </div>
                <div
                  class="token-line css-7zjwmr"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token plain"
                  >
                          
                  </span>
                  <span
                    class="token keyword"
                    style="color:rgb(86, 156, 214)"
                  >
                    const
                  </span>
                  <span
                    class="token plain"
                  >
                     b 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    =
                  </span>
                  <span
                    class="token plain"
                  >
                     
                  </span>
                  <span
                    class="token number"
                    style="color:rgb(181, 206, 168)"
                  >
                    2
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    ;
                  </span>
                  <span
                    class="token plain"
                  />
                </div>
                <div
                  class="token-line css-7zjwmr"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token plain"
                  >
                          
                  </span>
                  <span
                    class="token keyword"
                    style="color:rgb(86, 156, 214)"
                  >
                    const
                  </span>
                  <span
                    class="token plain"
                  >
                     c 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    =
                  </span>
                  <span
                    class="token plain"
                  >
                     a 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    *
                  </span>
                  <span
                    class="token plain"
                  >
                     a 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    +
                  </span>
                  <span
                    class="token plain"
                  >
                     b 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    *
                  </span>
                  <span
                    class="token plain"
                  >
                     b
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    ;
                  </span>
                  <span
                    class="token plain"
                  />
                </div>
                <div
                  class="token-line css-7zjwmr"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token plain"
                  >
                          
                  </span>
                  <span
                    class="token keyword"
                    style="color:rgb(86, 156, 214)"
                  >
                    return
                  </span>
                  <span
                    class="token plain"
                  >
                     
                  </span>
                  <span
                    class="token tag punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    &lt;
                  </span>
                  <span
                    class="token tag"
                    style="color:rgb(86, 156, 214)"
                  >
                    span
                  </span>
                  <span
                    class="token tag punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    &gt;
                  </span>
                  <span
                    class="token plain-text"
                  >
                    Answer: 
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    {
                  </span>
                  <span
                    class="token plain"
                  >
                    c
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    }
                  </span>
                  <span
                    class="token tag punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    &lt;/
                  </span>
                  <span
                    class="token tag"
                    style="color:rgb(86, 156, 214)"
                  >
                    span
                  </span>
                  <span
                    class="token tag punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    &gt;
                  </span>
                  <span
                    class="token plain"
                  />
                </div>
                <div
                  class="token-line css-7zjwmr"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token plain"
                  >
                        
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    }
                  </span>
                </div>
              </div>
            </div>
          </div>
        </pre>
      </div>
    </div>
    <div
      class="css-1wemi5a"
    >
      <div
        class="yozora-code-embed css-1lkb549"
      />
    </div>
  </div>
</div>
`;
